Basic Orchard feature tests

YevgeniyShunevych opened the pull request.

Piedone committed:
Placeholder for Orchard tests

YevgeniyShunevych committed:
- Add "Setup" and "Login" basic Orchard feature tests.
- Add and apply `SetsValueReliablyAttribute`.
- Refactor `OrchardCoreSetupPage`.

YevgeniyShunevych committed:
Add "Logout" basic Orchard feature test

Piedone commented on Lombiq.Tests.UI/Pages/OrchardCoreSetupParameters.cs:
```diff
@@ -9,8 +11,8 @@ public class OrchardCoreSetupParameters
-        public string Password { get; set; } = "Password1!";
+        public string UserName { get; set; } = DefaultUser.UserName;
+        public string Email { get; set; } = DefaultUser.UserName + "@admin.com";
+        public string Password { get; set; } = DefaultUser.Password;
```
Good point. While we're at it, we could have a const for the Email in `DefaultUser` too.

YevgeniyShunevych replied:
Added `Email` property.

Piedone committed:
Merge branch 'dev' into task/reusable-orchard-tests

Piedone committed:
Fixing that a setup snapshot was used even if the given test didn't need a snapshot

Piedone commented on Lombiq.Tests.UI/Attributes/Behaviors/SetsValueReliablyAttribute.cs:
```diff
@@ -0,0 +1,38 @@
+                // when entered into a text field so we need to use Actions. The following solution doesn't work:
+                // https://stackoverflow.com/a/52202594/220230. This needs to be done in addition to the standard
+                // FillInWith() as without that some forms start to behave strange and not save values.
+                AtataContext.Current.Driver.Perform(x => x.SendKeys(element, text));
```
Use a descriptive variable name instead of `x`.

YevgeniyShunevych replied:
Renamed.

Piedone commented on Lombiq.Tests.UI/Attributes/Behaviors/SetsValueReliablyAttribute.cs:
```diff
@@ -0,0 +1,38 @@
+{
+    public sealed class SetsValueReliablyAttribute : ValueSetBehaviorAttribute
+    {
+        public override void Execute<TOwner>(IUIComponent<TOwner> component, string value)
```
DRY this logic with the one in `FormUITestContextExtensions`.

YevgeniyShunevych replied:
There is a problem here. In `SetsValueReliablyAttribute` I don't have an instance of `UITestContext` in order to invoke its extension method.

Piedone replied:
You can just put the logic here into a third method that accepts the driver, and then use that from both places: Here it would get the driver from `AtataContext.Current`, otherwise from `UITestContext`.

YevgeniyShunevych replied:
OK. I was also thinking about that. Will do it.

YevgeniyShunevych replied:
Refactored.

YevgeniyShunevych committed:
Add `OrchardCoreFeaturesPage`.

YevgeniyShunevych committed:
Add `Email` property to `DefaultUser`

YevgeniyShunevych committed:
Refactor `SetsValueReliablyAttribute`

YevgeniyShunevych committed:
Extract `TryFillElement` method to `BasicWebDriverExtensions`

YevgeniyShunevych committed:
Add and update basic Orchard feature tests

- Add "Register" tests.
- Add "Turning features on and off" test.
- Refactor existing tests.

Piedone committed:
Using Atata shortcut for Actions

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -11,15 +12,22 @@ public static class BasicOrchardFeaturesTestingUITestContextExtensions
-            parameters ??= new OrchardCoreSetupParameters();
+            parameters ??= new OrchardCoreSetupParameters
+            {
+                RecipeId = "Lombiq.OSOCE.BasicOrchardFeaturesTests",
```
This can't really be a default as it's very specific to the OSOCE solution. It needs to be supplied from the sample code.

YevgeniyShunevych replied:
OK. So first of all I will then use `new OrchardCoreSetupParameters()` by default in `TestSetup` method. 

What if we add optional `OrchardCoreSetupParameters` parameter to `TestBasicOrchardFeatures`? Then I can update the test method this way:

```cs
[Theory, Chrome]
public Task BasicOrchardFeaturesShouldWork(Browser browser) =>
    ExecuteTestAsync(
        context => context.TestBasicOrchardFeatures(
            new OrchardCoreSetupParameters
            {
                RecipeId = "Lombiq.OSOCE.BasicOrchardFeaturesTests",
            }),
        browser);
```

Piedone replied:
Looks good!

YevgeniyShunevych replied:
Done,

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -63,6 +71,23 @@ public static UITestContext TestSetupNegatively(this UITestContext context, Orch
+                        .ShouldStayOnLoginPage()
+                        .ValidationSummaryErrors.Should.Not.BeEmpty();
+
+                    context.GetCurrentUserName().ShouldNotBe(userName);
```
```suggestion
                    context.GetCurrentUserName().ShouldBeEmpty();
```

YevgeniyShunevych replied:
Currently, how I see Orchard works. Being already logged in, user can go to login page and try to login. If login failed, the user remains still be logged in with previously login data. Current implementation covers this scenario.

So if we change assertion to `ShouldBeEmpty()` and invoke `TestLoginWithInvalidData` when user is logged in, we'll get an assertion exception `"should be empty but was "TestUser""`. So I see few solutions here:
1. Remain current implementation.
1. Use `ShouldBeEmpty()`.
1. Use `ShouldBeEmpty()`, but add logout step to the test beginning.

@Piedone, which one to choose?

Piedone replied:
A safety login step before with `SignOutDirectly()` is good.

YevgeniyShunevych replied:
Do you mean "A safety **logout** step"? Or something else?

Piedone replied:
Yes, sorry.

YevgeniyShunevych replied:
No problem. Updated the test.

Piedone commented on Lombiq.Tests.UI/Models/UserRegistrationModel.cs:
```diff
@@ -0,0 +1,21 @@
+
+namespace Lombiq.Tests.UI.Models
+{
+    public class UserRegistrationModel
```
```suggestion
    public class UserRegistrationParameters
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicWebDriverExtensions.cs:
```diff
@@ -0,0 +1,29 @@
+
+namespace Lombiq.Tests.UI.Extensions
+{
+    public static class BasicWebDriverExtensions
```
```suggestion
    public static class FormWebDriverExtensions
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -76,21 +101,78 @@ public static UITestContext TestLogout(this UITestContext context)
+                        .RegisterAsNewUser.Should.BeVisible();
 
-                    context.GetCurrentUserName().ShouldNotBe(userName);
+                    context.GoToRegistrationPage()
```
Let's click the previously checked link instead of going to the registration page directly.

YevgeniyShunevych replied:
Updated.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -76,21 +101,78 @@ public static UITestContext TestLogout(this UITestContext context)
+                context.GoToRegistrationPage()
+                    .RegisterWith(model)
+                    .ShouldStayOnRegistrationPage()
+                    .ValidationMessages[x => x.Email].Should.BeVisible());
```
Use a descriptive variable name instead of `x`.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -76,21 +101,78 @@ public static UITestContext TestLogout(this UITestContext context)
+        }
+
+        public static UITestContext TestTurningFeatureOnAndOff(this UITestContext context, string featureName = "Background Tasks")
+            =>
```
```suggestion
        public static UITestContext TestTurningFeatureOnAndOff(this UITestContext context, string featureName = "Background Tasks") =>
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -76,21 +101,78 @@ public static UITestContext TestLogout(this UITestContext context)
+                    .Features[featureName].CheckBox.Check()
+                    .BulkActions.Toggle.Click()
+
+                    .AggregateAssert(x => x
```
Use a descriptive variable name instead of `x`, also below.

Piedone commented on Lombiq.Tests.UI/Pages/OrchardCoreAdminPage.cs:
```diff
@@ -22,5 +23,8 @@ protected override void OnVerify()
         }
+
+        public TOwner ShouldContainSuccessAlertMessage(TermMatch expectedMatch, string expectedText) =>
+            AlertMessages.Should.Contain(x => x.IsSuccess && expectedMatch.IsMatch(x.Text.Value, expectedText));
```
Use a descriptive variable name instead of `x`.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -76,21 +101,78 @@ public static UITestContext TestLogout(this UITestContext context)
+                        .RegisterWith(model)
+                        .ShouldLeaveRegistrationPage();
+
+                    context.GetCurrentUserName().ShouldBe(model.UserName);
```
After this, it would be useful to log out. I'm not actually sure why `TestRegistrationWithAlreadyRegisteredEmail()` works after this but perhaps it's just an oversight that as a logged in user you can fill out the registration page.

YevgeniyShunevych replied:
Updated the test.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -0,0 +1,188 @@
+                () =>
+                context.GoToSetupPage()
+                    .SetupOrchardCore(parameters)
+                    .ShouldLeaveSetupPage());
```
Similarly in the other cases too:

```suggestion
                () => context
                    .GoToSetupPage()
                    .SetupOrchardCore(parameters)
                    .ShouldLeaveSetupPage());
```

YevgeniyShunevych committed:
Add "Content operations" basic Orchard feature test

YevgeniyShunevych committed:
Refactor test classes

YevgeniyShunevych committed:
Rename `BasicWebDriverExtensions` to `FormWebDriverExtensions`

YevgeniyShunevych committed:
Rename `UserRegistrationModel` to `UserRegistrationParameters`

YevgeniyShunevych committed:
Update `TestRegistration` method

YevgeniyShunevych committed:
Add `OrchardCoreSetupParameters` parameter to `TestBasicOrchardFeatures` method

YevgeniyShunevych committed:
Update `TestLoginWithInvalidData`

YevgeniyShunevych committed:
Merge branch 'dev' into task/reusable-orchard-tests

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -132,44 +137,65 @@ public static UITestContext TestRegistrationWithInvalidData(this UITestContext c
 
-        public static UITestContext TestTurningFeatureOnAndOff(this UITestContext context, string featureName = "Background Tasks")
+        public static UITestContext TestContentOperations(this UITestContext context, string pageTitle = "Test page")
             =>
```
There are a couple of other such examples in this class too.

```suggestion
        public static UITestContext TestContentOperations(this UITestContext context, string pageTitle = "Test page") =>
```

YevgeniyShunevych replied:
Updated.

Piedone commented on Lombiq.Tests.UI/Attributes/Behaviors/SetsValueReliablyAttribute.cs:
```diff
@@ -0,0 +1,18 @@
+        public override void Execute<TOwner>(IUIComponent<TOwner> component, string value)
+        {
+            var element = component.Scope;
+            var driver = AtataContext.Current.Driver;
```
There are a couple of such `AtataContext.Current` references. We've consciously avoided this until now due to the references possibly being lost when `async` code is in the tests, especially if multiple ones are executed in parallel. Hence `UITestContext.Scope.AtataContext` too. Please use `UITestContext` instead too.

YevgeniyShunevych replied:
I thought that `AtataContext.Current` should not be such issue as before, as there was improvement on that. Anyway, `AtataContext.Current` is used in Atata itself in some places.

I would like to use `UITestContext` here but I can't get a reference to current instance in behavior attribute.

Either way, I have a suggestion. I can make [Driver property of UIComponent](https://github.com/atata-framework/atata/blob/bf4cbd0dd2974c2b5112f3ee6107e129d154bccb/src/Atata/Components/UIComponent.cs#L31) public and add it to `IUIComponent<TOwner>`. Then instead of using `AtataContext.Current.Driver` we'll be able to use `component.Driver`. I'll just need a couple of days to release Atata v1.13.0 with this change, as there are few other things that I want to add to upcoming Atata version release.

Piedone replied:
Even after the improvements we had issues, hence the `GoToPage<T>()` extension to avoid using `AtataContext.Current`.

While using `component.Driver` would work here, we also have `AtataContext.Current.BaseUrl` references. Why not just add a `UITestContext` property onto the page classes that you can then access?

YevgeniyShunevych replied:
I thought and reviewed Atata a bit. I am thinking to add `public AtataContext Context { get; }` property to `UIComponent`. This way we'll have and be able to use `AtataContext` instance in page objects, in attributes, etc.

Piedone replied:
Would that mean that after `UITestContext.Scope.AtataContext.Go.To<MyPage>()` you'd be able to access the same `UITestContext.Scope.AtataContext` (and not `AtataContext.Current`) from inside `MyPage`? If yes then great!

YevgeniyShunevych replied:
Basically, yes.

Piedone replied:
Sounds good then.

YevgeniyShunevych replied:
I've released Atata v1.13.0 and updated `AtataContext` related code here.

Piedone committed:
Docs on basic OC feature tests

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -0,0 +1,214 @@
+{
+    public static class BasicOrchardFeaturesTestingUITestContextExtensions
+    {
+        public static UITestContext TestBasicOrchardFeatures(this UITestContext context, OrchardCoreSetupParameters setupParameters = null) =>
```
Add some docs on the public methods, please, especially outlining any assumptions they make, like the site not being set up.

YevgeniyShunevych replied:
I've added XML comments. Please check it out.

YevgeniyShunevych committed:
Refactor `=>` in `BasicOrchardFeaturesTestingUITestContextExtensions`

YevgeniyShunevych committed:
Add XML comments to `BasicOrchardFeaturesTestingUITestContextExtensions`

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -8,8 +8,25 @@
+        /// Tests all the basic Orchard features.
+        /// At first sets up Orchard with optionally specified <paramref name="setupParameters"/>.
+        /// By default uses new <see cref="OrchardCoreSetupParameters"/> instance
+        /// with <c>"SaaS"</c> <see cref="OrchardCoreSetupParameters.RecipeId"/> value.
```
If you want multiple paragraphs here (it seems to me based on the line breaks) then add multiple `<para>` tags, please. Otherwise, break comments at 120 characters (i.e. the first soft limit also added by the Analyzers). This extension can do this automatically: https://marketplace.visualstudio.com/items?itemName=stkb.rewrap

YevgeniyShunevych replied:
OK, reformatted.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -8,8 +8,25 @@
+        /// with <c>"SaaS"</c> <see cref="OrchardCoreSetupParameters.RecipeId"/> value.
+        /// </para>
+        /// <para>
+        /// The test method assumes that site is not set up.
```
Also below:

```suggestion
        /// The test method assumes that the site is not set up.
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -8,8 +8,25 @@
+        /// The test method assumes that site is not set up.
+        /// </para>
+        /// </summary>
+        /// <param name="context">The context.</param>
```
No need to add a comment for the `this` parameter as you never have to explicitly provide it.

YevgeniyShunevych replied:
OK, removed. I just get used to have a StyleCop warning in Atata for missing parameter comment.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -35,6 +65,21 @@ public static UITestContext TestSetup(this UITestContext context, OrchardCoreSet
 
+        /// <summary>
+        /// <para>
+        /// Tests negatively the site setup with optionally set <paramref name="parameters"/>.
```
```suggestion
        /// Tests the site setup negatively with optionally set <paramref name="parameters"/>.
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -101,6 +177,19 @@ public static UITestContext TestLogout(this UITestContext context)
+        /// After the user is registered, the test performs login with the user credentials, then logout.
+        /// </para>
+        /// <para>
+        /// The test method assumes that "Users Registration" Orchard feature is enabled
```
```suggestion
        /// The test method assumes that the "Users Registration" Orchard feature is enabled
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -101,6 +177,19 @@ public static UITestContext TestLogout(this UITestContext context)
+        /// </para>
+        /// <para>
+        /// The test method assumes that "Users Registration" Orchard feature is enabled
+        /// and there is no registered user with the given <see cref="UserRegistrationParameters.Email"/> value.
```
Should be email or username values, or isn't this how it works?

YevgeniyShunevych replied:
You are right. I've updated the comment.

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -125,6 +214,19 @@ public static UITestContext TestRegistration(this UITestContext context, UserReg
 
+        /// <summary>
+        /// <para>
+        /// Tests negatively the user registration with optionally specified invalid <paramref name="parameters"/>.
```
```suggestion
        /// Tests the user registration negatively with optionally specified invalid <paramref name="parameters"/>.
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -125,6 +214,19 @@ public static UITestContext TestRegistration(this UITestContext context, UserReg
+        /// clicks "Register" button and verifies that there are validation messages on the page.
+        /// </para>
+        /// <para>
+        /// The test method assumes that "Users Registration" Orchard feature is enabled.
```
```suggestion
        /// The test method assumes that the "Users Registration" Orchard feature is enabled.
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -144,6 +246,21 @@ public static UITestContext TestRegistrationWithInvalidData(this UITestContext c
 
+        /// <summary>
+        /// <para>
+        /// Tests negatively the user registration with optionally specified <paramref name="parameters"/>
```
```suggestion
        /// Tests the user registration negatively with optionally specified <paramref name="parameters"/>
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -144,6 +246,21 @@ public static UITestContext TestRegistrationWithInvalidData(this UITestContext c
+        /// clicks "Register" button and verifies that there is a validation message near "Email" field on the page.
+        /// </para>
+        /// <para>
+        /// The test method assumes that "Users Registration" Orchard feature is enabled
```
```suggestion
        /// The test method assumes that the "Users Registration" Orchard feature is enabled
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -144,6 +246,21 @@ public static UITestContext TestRegistrationWithInvalidData(this UITestContext c
+        /// </para>
+        /// <para>
+        /// The test method assumes that "Users Registration" Orchard feature is enabled
+        /// and there is already registered user with the given <see cref="UserRegistrationParameters.Email"/> value.
```
```suggestion
        /// and there is an already registered user with the given <see cref="UserRegistrationParameters.Email"/> value.
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -159,8 +276,26 @@ public static UITestContext TestRegistrationWithInvalidData(this UITestContext c
+        /// Tests content operations. The test executes the following steps:
+        /// </para>
+        /// <list type="number">
+        /// <item><description>Navigate to "Content / Content Items" page.</description></item>
```
```suggestion
        /// <item><description>Navigate to the "Content / Content Items" page.</description></item>
```

Piedone commented on Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs:
```diff
@@ -179,6 +314,26 @@ public static UITestContext TestContentOperations(this UITestContext context, st
+        /// Tests turning feature on and off. The test executes the following steps:
+        /// </para>
+        /// <list type="number">
+        /// <item><description>Navigate to "Configuration / Features" page.</description></item>
```
```suggestion
        /// <item><description>Navigate to the "Configuration / Features" page.</description></item>
```

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update XML comments of `BasicOrchardFeaturesTestingUITestContextExtensions`

YevgeniyShunevych committed:
Update XML comments of `BasicOrchardFeaturesTestingUITestContextExtensions`…

YevgeniyShunevych committed:
Update XML comments of `BasicOrchardFeaturesTestingUITestContextExtensions`

YevgeniyShunevych committed:
Update Lombiq.Tests.UI/Extensions/BasicOrchardFeaturesTestingUITestContextExtensions.cs

Co-authored-by: Zoltán Lehóczky <zoltan.lehoczky@lombiq.com>

YevgeniyShunevych committed:
Update XML comments of `BasicOrchardFeaturesTestingUITestContextExtensions`

YevgeniyShunevych committed:
Update XML comments of `BasicOrchardFeaturesTestingUITestContextExtensions`

YevgeniyShunevych committed:
Upgrade Atata package to v1.13.0. Apply few latest Atata features.

YevgeniyShunevych committed:
Merge branch 'dev' into task/reusable-orchard-tests

Piedone committed:
Updating note about AtataContext.Current
